mproton_darwin.m

```diff
@@ -0,0 +1,255 @@
+                 replyHandler:(nonnull void (^)(id _Nullable, NSString * _Nullable))replyHandler;
+@end
+
+@implementation AppContext
+- (void)userContentController:(nonnull WKUserContentController *)userContentController
+      didReceiveScriptMessage:(nonnull WKScriptMessage *)message
+                 replyHandler:(nonnull void (^)(id _Nullable, NSString * _Nullable))replyHandler {
+    
+    NSLog(@"[objc] WKScriptMessageHandlerWithReply callback called");
+    
+    extern const char * goCallbackDispatcher(const void * _Nonnull, const char * _Nonnull);
+    
+    dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0), ^{
+        struct goTrampoline_return result = goTrampoline(17, 
+			(char*)([message.name UTF8String]), 
+			(char*)([message.body UTF8String]));
+        
+        NSString * r0 = result.r0 != NULL ? [NSString stringWithUTF8String:result.r0] : NULL;
+        NSString * r1 = result.r1 != NULL ? [NSString stringWithUTF8String:result.r1] : NULL;
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
+			replyHandler(r0, r1);
+        });
+    });
+}
+@end
+
+AppContext * g_appContext = nil;
+
+
+@interface AppDelegate : NSObject <
+NSApplicationDelegate
+>
+
+@property(nonatomic, retain)  NSWindow  * _Nullable window;
+@property(nonatomic, retain)  NSMenu  * _Nullable mainMenu;
+@property(nonatomic, retain)  NSStatusItem * _Nullable statusItem;
+
+- (void)applicationWillFinishLaunching:(NSNotification *)notification;
+- (void)applicationDidFinishLaunching:(NSNotification *)notification;
+@end
+
+
+@implementation AppDelegate
+
+- (BOOL)applicationShouldTerminateAfterLastWindowClosed:(NSApplication *)sender {
+    NSLog(@"in applicationShouldTerminateAfterLastWindowClosed");
+
+	[NSApp stop:nil];   
+    
+	return FALSE;
+}
+
+static NSWindow * createWindow(id appName) {
+    NSUInteger style = NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable |  NSWindowStyleMaskResizable;
+    NSWindow *window = [[NSWindow alloc]
+                        initWithContentRect:NSMakeRect(0, 0, 800, 600)
+                        styleMask:style
+                        backing:NSBackingStoreBuffered
+                        defer:NO];
+    [window cascadeTopLeftFromPoint:NSMakePoint(20,20)];
+    [window setTitle:appName];
+    [window makeKeyAndOrderFront:nil];
+    
+    return window;
+}
+
+static NSMenu * createMainMenu(id appName) {
+    id appMenu = [[NSMenu alloc] init];
+    [appMenu addItem:[[NSMenuItem alloc]
+                      initWithTitle:[@"Quit " stringByAppendingString:appName]
+                      action:@selector(terminate:)
+                      keyEquivalent:@"q"]];
+    
+    id appMenuItem = [[NSMenuItem alloc] init];
+    [appMenuItem setSubmenu:appMenu];
+    
+    NSMenu *mainMenu = [[NSMenu alloc] init];
+    [mainMenu addItem:appMenuItem];
+    
+    return mainMenu;
+}
+
+static NSStatusItem * createMenuExtra() {
```

Comment:
"Menu Extra" is kind of confusing. Maybe call these tray or system status?

Reply:
Funny topic, "Extra" term comes from MacOS [1], interestingly enough GTK has dropped the concept all together ... 

[1] https://developer.apple.com/design/human-interface-guidelines/macos/extensions/menu-bar-extras/